### PR TITLE
Create root folder and declare materials sidebar item

### PR DESCRIPTION
**Pull request by allenwq**

### CODE REVIEW

File: app/controllers/components/course/materials_component.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class Course::MaterialsComponent < SimpleDelegator
+  include Course::ControllerComponentHost::Component
+
+  def sidebar_items
+    [
+      {
+        key: :materials,
+        title: I18n.t('course.material.sidebar_title'),
+        weight: 4,
+        path: course_material_folder_path(current_course, current_course.root_folder),
```
Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: @lowjoel We'll need to create root folders for all existing courses on staging server after this was merged

File: app/models/course.rb [hunk 2]
```diff
@@ -35,7 +35,8 @@ class Course < ActiveRecord::Base
                                 source: :actable, source_type: Course::LessonPlan::Event.name
   has_many :material_folders, class_name: Course::Material::Folder.name, inverse_of: :course,
                               dependent: :destroy
-
+  has_one :root_folder, -> { where(parent_id: nil) },
```
Comment (allenwq) [inline app/models/course.rb]: @lowjoel there is an alternative solution to resolve the issue above:

```
def root_folder
   material_folders.first || create_root_folder
end
```

It's more reliable, it ensures that every course will return a root folder. But the implementation which I push is more common, what do you think ?

Comment (allenwq) [inline app/models/course.rb]: ...And we might need a conditional unique index for current implementation as well, since consistency_fail failed.

File: app/models/course.rb [hunk 3]
```diff
@@ -53,6 +53,12 @@ def generate_registration_key
     self.registration_key = 'C'.freeze + SecureRandom.base64(8)
   end
 
+  # Returns the root folder of the course.
+  # @return [Course::Material::Folder] The root folder.
+  def root_folder
+    material_folders.find_by(parent: nil)
```
Comment (lowjoel) [inline app/models/course.rb]: should this be `find_by!`

File: app/models/course/material/folder.rb [hunk 4]
```diff
@@ -1,7 +1,9 @@
 class Course::Material::Folder < ActiveRecord::Base
   acts_as_forest order: 'name'
+  include Course::ModelComponentHost::Component
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: Do we include this in other models?

File: app/models/course/material/folder.rb [hunk 5]
```diff
@@ -1,7 +1,9 @@
...
 
   after_initialize :set_defaults, if: :new_record?
+  before_destroy :destroy_children
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: does edge deal with this?

File: app/models/course/material/folder.rb [hunk 6]
```diff
@@ -10,9 +12,17 @@ class Course::Material::Folder < ActiveRecord::Base
   # TODO: Remove this after schema_validations #21 was closed
   validates :name, uniqueness: { case_sensitive: false, scope: :parent_id }, if: :parent
 
+  def self.after_course_create(course)
+    course.material_folders.create(name: 'Root') unless course.root_folder
```
Comment (lowjoel) [inline app/models/course/material/folder.rb]: localisation?

Comment (lowjoel) [inline app/models/course/material/folder.rb]: oh, wait, this won't ever be seen by the user...

Comment (lowjoel) [inline app/models/course/material/folder.rb]: why the unless clause?

File: app/controllers/components/course/materials_component.rb (hunk 1 above)
Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: probably should be part of your migration.

File: app/models/course.rb (hunk 3 above)
Comment (allenwq) [inline app/models/course.rb]: Ok, so you want raise an exception when there is no ?

File: app/models/course/material/folder.rb (hunk 5 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: No, that's why I add it here.

Comment (lowjoel) [inline app/models/course/material/folder.rb]: you might be able to get away with acts_as_forest order: :name, dependent: :destroy

File: app/models/course.rb (hunk 3 above)
Comment (lowjoel) [inline app/models/course.rb]: that's a problem with the course if we don't have this right? Otherwise the nil will probagate downwards.

File: app/models/course/material/folder.rb (hunk 4 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: Yes, I'm doing the same thing as assessment category

File: app/models/course/material/folder.rb (hunk 5 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: Oh yes,  that might work, thanks !

File: app/models/course/material/folder.rb (hunk 6 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: Oops. I think it's not needed anymore, it's for solving the validation issue before.

File: app/controllers/components/course/materials_component.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: Migration ? sorry don't get you..

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: you need to ensure all the courses have a root folder right?

File: app/models/course/material/folder.rb (hunk 4 above)
Comment (lowjoel) [inline app/models/course/material/folder.rb]: hmmmm this could be improved but hmmm okay.

File: app/controllers/components/course/materials_component.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: Yes, but how to do that through migration ?  You mean rake tasks ?

File: app/models/course/material/folder.rb (hunk 5 above)
Comment (allenwq) [inline app/models/course/material/folder.rb]: @lowjoel Unfortunately, edge does not support dependent options :(
https://github.com/jackc/edge/blob/master/lib/edge/forest.rb#L10

File: app/controllers/components/course/materials_component.rb (hunk 1 above)
Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: Migrations take a schema and transforms them. You should not use models within the migration, but executing insert statements in the up method is okay.

File: app/models/course/material/folder.rb (hunk 5 above)
Comment (lowjoel) [inline app/models/course/material/folder.rb]: I'd still prefer to use order: :name instead of order: 'name'.

File: app/controllers/components/course/materials_component.rb (hunk 1 above)
Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: But why not use rake script ?  And there is no migration file in this PR.

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: because then we get a collection of rake scripts that mean nothing.

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: you can always create a migration.

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: rake tasks are supposed to be reusable, things like these are single-use only.

Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: OK I agree that rake task should be reusable.

But my another question is why " should not use models within the migration" ?  If not using model in migrations, means we skipped the activerecord's abstraction, things will easily go wrong.

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: Google around. Many answers.

In short: migration can be run after the model changes. Then your migration is broken.

Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: But if you directly update at db level, manually manipulating all those fields (creator, updater, updated_at, created_at, etc), it still very possible that you will break things.

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: So try to think ahead. I don't have a solution to this.

Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: To think further, if changes are complicated, we probably will use Models, otherwise I don't think we can make it.

So I would still prefer using a rake task, but not commit to git, They are mostly temporary tasks. new deployment won't need to run such tasks.

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: This is not going to be a problem that will not recur. You will have changes to the schema and data even after deployment. No point trying to run away from it.

I already said you cannot use models.

Comment (allenwq) [inline app/controllers/components/course/materials_component.rb]: If changes are in the schema, we will use migration scripts.

But this is not related to schema. 

I understand, I mean using models in rake tasks. not migration.

Comment (lowjoel) [inline app/controllers/components/course/materials_component.rb]: If you use the models in the rake tasks, you have the exact same problem as using it in the migration script, you know?

**Comment by allenwq** [comment]
@lowjoel I added a migration script for creating folders for existing courses.
